File: src/02-rest-api/typescript/src/main.ts
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
 */

import express, { Request, Response } from "express";

const app = express();

const books: string[] = [
  "Jane Austen: Pride and Prejudice",
  "George Orwell: 1984",
  "F. Scott Fitzgerald: The Great Gatsby",
];

app.get("/books", (req: Request, res: Response) => {
  const booksObject = { books: books };
  res.json(booksObject);
});

const PORT: number = process.env.PORT ? parseInt(process.env.PORT) : 3000;
app.listen(PORT, () => {
  console.log("Rest API example");
  console.log(`Server is running on port ${PORT}`);
  console.log();
  console.log("Try the following: GET /books");
});

File: src/02-rest-api/rust/src/main.rs
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
 */

use axum::{routing::get, Json, Router};
use serde::Serialize;

#[derive(Serialize)]
struct MyBooks {
    books: Vec<String>,
}

async fn get_books() -> Json<MyBooks> {
    let books: Vec<String> = [
        "Jane Austen: Pride and Prejudice",
        "George Orwell: 1984",
        "F. Scott Fitzgerald: The Great Gatsby",
    ]
    .iter()
    .map(|&s| s.to_string())
    .collect();

    let my_books: MyBooks = MyBooks { books };

    Json(my_books)
}

#[tokio::main(flavor = "current_thread")]
async fn main() {
    println!("Rest API example");
    println!("Server is running on 0.0.0.0:3000");
    println!();
    println!("Try the following: GET /books");

    let app = Router::new().route("/books", get(get_books));

    let listener = tokio::net::TcpListener::bind("0.0.0.0:3000").await;

    match listener {
        Ok(ok_listener) => {
            println!("Listener is working...");
            let serve = axum::serve(ok_listener, app).await;
            match serve {
                Ok(_ok_serve) => {
                    println!("Serving is working...");
                }
                Err(e) => {
                    println!("Error when serving server. {}", e);
                }
            }
        }
        Err(e) => {
            println!("Error when binding to TcpListener. {}", e);
        }
    }
}
